packages/webpack/src/plugin.js

```diff
@@ -0,0 +1,362 @@
+          PROGRESS.report('pathsCollected')
+
+          identifierLookup = generateIdentifierLookup({
+            readableResourceIds: options.readableResourceIds,
+            unenforceableModuleIds,
+            paths: knownPaths,
+            policy: options.policy,
+            canonicalNameMap,
+          })
+          if (unenforceableModuleIds.length > 0) {
+            mainCompilationWarnings.push(
+              new WebpackError(
+                `LavaMoatPlugin: the following module ids can't be controlled by policy and must be ignored at runtime: \n  ${unenforceableModuleIds.join()}`,
+              ),
+            )
+          }
+          PROGRESS.report('pathsProcessed')
+        })
+        // lists modules, but reaching for id can give a deprecation warning.
+        // compilation.hooks.afterOptimizeModuleIds.tap(PLUGIN_NAME, (modules) => {
+        //   console.log('________________________________',modules.__proto__.constructor)
+        //   Array.from(modules).map(module => console.log(module.resource, module.id))
+        // });
+
+        // =================================================================
+        // javascript modules generator tweaks installation
+
+        for (const moduleType of coveredTypes) {
+          normalModuleFactory.hooks.generator.for(moduleType).tap(
+            PLUGIN_NAME,
+            wrapGeneratorMaker({
+              ignores,
+              runChecks,
+              getIdentifierForPath,
+              PROGRESS,
+            }),
+          )
+        }
+
+        // Report on ignored modules as late as possible.
+        // This hook happens after all module generators have been executed.
+        compilation.hooks.afterProcessAssets.tap(PLUGIN_NAME, () => {
+          diag.rawDebug(3, '> afterProcessAssets')
+          mainCompilationWarnings.push(
+            new WebpackError(
+              `in LavaMoatPlugin: ignored modules \n  ${ignores.join('\n  ')}`,
+            ),
+          )
+        })
+
+        // =================================================================
+
+        // This part adds LavaMoat runtime to webpack runtime for every chunk that needs runtime.
+        // I stole this from Zach of module federation fame
+
+        const onceForChunkSet = new WeakSet()
+        const runtimeOptions = {
+          lockdown: options.lockdown,
+        }
+
+        // Define a handler function to be called for each chunk in the compilation.
+        compilation.hooks.additionalChunkRuntimeRequirements.tap(
+          PLUGIN_NAME + '_runtime',
+          (chunk /*, set*/) => {
+            if (chunk.hasRuntime()) {
+              if (!PROGRESS.done('gneratorCalled')) {
+                mainCompilationWarnings.push(
+                  new WebpackError(
+                    'LavaMoatPlugin: Something was generating runtime before all modules were identified. This might be part of a sub-compilation of a plugin. Please check for any unwanted interference between plugins.',
```

Comment:
I'm curious if this is only a hazard in our case or in the general case

Reply:
It's gonna be helpful for interactions with come weird plugins when stuff goes sideways. 
Also, once we get a better understanding of how it's used and when, we might actually prevent webpack from running code form the frontend bundle at build time (it seems it's possible for that to happen. Imagine a styled component stealing your github token...)